---
title: "Univariate tests"
author: "`r Sys.Date()`"
bibliography: 
- bibliography.bib
output:
  rmdformats::readthedown:
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: true
    use_bookdown: false
    highlight: haddock
---
<!--
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{microbiome tutorial - atlas}
  %\usepackage[utf8]{inputenc}
-->



```{r, echo=FALSE, message=FALSE, warning=FALSE}
# Handle citations
require(knitcitations)
require(bookdown)
# cleanbib()
# options("citation_format" = "pandoc")
bib <- read.bibtex("bibliography.bib")
#opts_chunk$set(fig.width=4, fig.height=3, par=TRUE, out.width='2in', fig.pos='H')
library(knitr)
knitr::opts_chunk$set(fig.path = "figure/", dev="CairoPNG")
```


# Differential abundance testing for univariate data

This section covers basic univariate tests for two-group comparison,
covering t-test, Wilcoxon test, and multiple testing.
  
The following example compares the abundance of a selected bug between
two conditions. Let us assume that the data is already properly
normalized.

Let us load exmaple data

```{r univariate1, warning=FALSE, message=FALSE}
library(microbiome)
data(dietswap)
d <- dietswap

# Pick microbial abundances for a given taxonomic group
taxa <- "Dialister"

# Construct a data.frame with the selected
# taxonomic group and grouping
df <- data.frame(Abundance = abundances(d)[taxa,],
                 Group = meta(d)$nationality)

```


Compare the groups visually using a boxplot.

```{r univariate_boxplot, warning=FALSE, message=FALSE}
p <- ggplot(df, aes(x = Group, y = Abundance)) +
       geom_boxplot()
print(p)
```



The abundances are in absolute scale and comparison is not clear.
Therefore, let us use the log10 transformation.
Now, the data contains many zeros and taking log10 will yield
infinite values. Let us use the standard log10(1+x) transformation.

```{r univariate2, warning=FALSE, message=FALSE}
df$Log10_Abundance <- log10(1 + df$Abundance)
p <- ggplot(df, aes(x = Group, y = Log10_Abundance)) +
       geom_boxplot()
print(p)
```



The groups seems to differ. Let us test the difference statistically.
First, let us perform t-test, which is based on Gaussian assumptions.
Each group is expected to follow Gaussian distribution.

Significance with t-test:

```{r univariate2b, warning=FALSE, message=FALSE}
print(t.test(Log10_Abundance ~ Group, data = df)$p.value)
```


Now let us investigate the Gaussian assumption in more
detail. Boxplots may not show deviations from Gaussian assumptions
very clearly Let us try another visualization; the density plot.

```{r univariate_densityplot, warning=FALSE, message=FALSE}
p <- ggplot(df, aes(fill = Group, x = Log10_Abundance)) +
       geom_density(alpha = 0.5)
print(p)
```


The data is not Gaussian distributed.  In such cases, a common
procedure is to use non-parametric tests. These do not make
assumptions of the data distribution but instead compare the ordering
of the samples.


Significance with Wilcoxon test (log10 data):

```{r univariate_wilcoxon, warning=FALSE, message=FALSE}
print(wilcox.test(Log10_Abundance ~ Group, data = df)$p.value)
```


However, since the test is non-parametric, we do not ned need
log10(1+x) transformation since this is a monotonous transformation
that does not affect sample ordering. Let us verify that the absolute
abundances yield the same p-value:

Significance with Wilcoxon test (absolute data:

```{r univariate_wilcoxon2, warning=FALSE, message=FALSE}
print(wilcox.test(Abundance ~ Group, data = df)$p.value)
```


Let us compare how much the results would differ in the whole data Let
us only invetigate core taxa that are observed in more than 20% of the
samples with more than 3 reads.

```{r univariate4, warning=FALSE, message=FALSE}
all.taxa <- core_members(d, prevalence = 20/100, detection = 3)

pvalue.ttest <- c()
pvalue.wilcoxon <- c()
for (taxa in all.taxa) {
  # Create a new data frame for each taxonomic group
  df <- data.frame(Abundance = abundances(d)[taxa,],
                   Log10_Abundance = log10(1 + abundances(d)[taxa,]),  
                   Group = meta(d)$nationality)

  pvalue.ttest[[taxa]] <- t.test(Log10_Abundance ~ Group, data = df)$p.value
  pvalue.wilcoxon[[taxa]] <- wilcox.test(Abundance ~ Group, data = df)$p.value  
}
pvalues <- data.frame(taxon = all.taxa,
	              pvalue.ttest = pvalue.ttest,
                      pvalue.wilcoxon = pvalue.wilcoxon)

# Now, note that multiple testing occurs.
# We must correct the p-values.
# let us apply the standard Benjamini-Hochberg False Discovery Rate (FDR)
# correction
pvalues$pvalue.ttest.adjusted <- p.adjust(pvalues$pvalue.ttest)
pvalues$pvalue.wilcoxon.adjusted <- p.adjust(pvalues$pvalue.wilcoxon)
```


Compare the p-value histograms between raw and adjusteed p-values.

```{r univariate5, warning=FALSE, message=FALSE}
p1 <- ggplot(pvalues, aes(x = pvalue.wilcoxon)) +
        geom_histogram() +
        labs(title = "Raw p-values") #+
	#ylim(c(0, 80))

p2 <- ggplot(pvalues, aes(x = pvalue.wilcoxon.adjusted)) +
        geom_histogram() +
        labs(title = "Adjusted p-values") #+
	#ylim(c(0, 80))	

library(patchwork)
print(p1 + p2)
```


Compare adjusted p-values between t-test and Wilcoxon test. Let us
also highlight the p = 0.05 intervals.

```{r univariate6, warning=FALSE, message=FALSE}
p <- ggplot(data = pvalues,
         aes(x = pvalue.ttest.adjusted,
	     y = pvalue.wilcoxon.adjusted)) +
       geom_text(aes(label = taxon)) + 
       geom_abline(aes(intercept = 0, slope = 1)) +
       geom_hline(aes(yintercept = 0.05)) +
       geom_vline(aes(xintercept = 0.05)) #+
       #scale_x_log10() +
       #scale_y_log10()        
print(p)
```













